### PR TITLE
Provision available mailboxes for Professional Email

### PR DESCRIPTION
**Pull request by olaseni**
#### Changes proposed in this Pull Request

This PR adds a new page i.e.  `/email/<domain>/titan/setup-mailbox/<site-slug>` which contains a form for creating a  mailbox as an additional post-signup task for Professional Email opt-ins.

- The form accepts mailbox parameters and sends them via a REST API for provisioning on the backend.
- Adds a parameter to the existing `TitanNewMailboxList` component to support adding a single mailbox
- Adds two react query components, a query that performs mailbox availability lookups, and a mutation that actually creates the mailbox resource if it's available.

This is part of the new effort to add Professional Email as part of the sign-up flow.

*

#### Testing instructions

- Checkout the branch, or use the [live branch](https://calypso.live/?image=registry.a8c.com/calypso/app:build-11347)
- Select a site where you have an existing Professional Email subscription
- Navigate to `/email/<domain>/titan/setup-mailbox/<site-slug>` 
- Confirm that the form is visible
- Confirm that you are redirected to the Email Upsell page when there's no Titan subscription
- Attempt to create a mailbox with an existing name. Confirm that a conflict error is returned
- Attempt to create a mailbox without any free licenses/slots. Confirm that an appropriate error is returned.
- Confirm that a mailbox can be created when all conditions are favourable
- Confirm a redirect to the Email Plan page when successful. NOTE: This is temporary until the Thank You page is ready.

Screenshot 2021-07-27 at 6 30 58 PM

### CODE REVIEW

File: client/my-sites/email/titan-complete-mailbox-setup/index.js
Comment (a8ci18n) [inline client/my-sites/email/titan-complete-mailbox-setup/index.js]: :information_source: String reuse speeds up translation and improves consistency. The following string might make a good alternative and has already been translated **31** times:
`translate( 'Complete setup' )` *ES Score: 10*

File: client/my-sites/email/titan-add-mailboxes/titan-new-mailbox-list.jsx
Comment (daledupreez) [inline client/my-sites/email/titan-add-mailboxes/titan-new-mailbox-list.jsx]: Nit: I would prefer `allowMultipleMailboxes` as it captures the directive we're giving the component. I feel `supportsMultipleMailboxes` is less clear and it's not immediately clear to me what that might mean for the component.

File: client/my-sites/email/titan-complete-mailbox-setup/index.js
Comment (daledupreez) [inline client/my-sites/email/titan-complete-mailbox-setup/index.js]: Shouldn't this be `mailbox`?

File: client/my-sites/email/titan-complete-mailbox-setup/index.js
Comment (daledupreez) [inline client/my-sites/email/titan-complete-mailbox-setup/index.js]: Shouldn't this also fire in the case where `mailboxIsValid` is false and `isError` is true? i.e. just before the `return;` on L106.

File: client/my-sites/email/titan-complete-mailbox-setup/index.js
Comment (daledupreez) [inline client/my-sites/email/titan-complete-mailbox-setup/index.js]: This `<>` wrapper seems unnecessary - the only child element is a single `Card` element.

File: client/my-sites/email/titan-complete-mailbox-setup/index.js
Comment (daledupreez) [reply client/my-sites/email/titan-complete-mailbox-setup/index.js]: This may be worth using the existing string!

File: client/my-sites/email/titan-complete-mailbox-setup/index.js
Comment (daledupreez) [inline client/my-sites/email/titan-complete-mailbox-setup/index.js]: I think we can use these two functions directly -- they are acting directly against the supplied arguments and don't depend on `state`.

File: client/my-sites/email/titan-complete-mailbox-setup/index.js
Comment (daledupreez) [inline client/my-sites/email/titan-complete-mailbox-setup/index.js]: Maybe call this `"Email Management > Set up your Professional Email mailbox"` to be slightly more explicit?
And please change "Setup" to "Set up"!

File: client/my-sites/email/titan-complete-mailbox-setup/index.js
Comment (daledupreez) [inline client/my-sites/email/titan-complete-mailbox-setup/index.js]: As above, I think this should be "Set up your Professional Email".

**Review by daledupreez — COMMENTED** [review]
Here are my comments from code inspection only -- it looks like there may be one or two bugs (or places that need comments!).

File: client/my-sites/email/titan-add-mailboxes/titan-new-mailbox-list.jsx
Comment (olaseni) [reply client/my-sites/email/titan-add-mailboxes/titan-new-mailbox-list.jsx]: 🤔  I feel `supportsMultipleMailboxes` depicts this better. 

In computing we often say a feature is **supported** if that feature is accessible/capable to be used/managed. 

Most times if a feature is **allowed** , it may be related to permission/authorisation. Better put, a feature may be supported, but not allowed.

File: client/my-sites/email/titan-complete-mailbox-setup/index.js
Comment (olaseni) [reply client/my-sites/email/titan-complete-mailbox-setup/index.js]: Good catch. I tripped over some last minute refactors.

File: client/my-sites/email/titan-complete-mailbox-setup/index.js
Comment (olaseni) [reply client/my-sites/email/titan-complete-mailbox-setup/index.js]: 👍

File: client/my-sites/email/titan-complete-mailbox-setup/index.js
Comment (olaseni) [reply client/my-sites/email/titan-complete-mailbox-setup/index.js]: Good catch!

File: client/my-sites/email/titan-complete-mailbox-setup/index.js
Comment (olaseni) [reply client/my-sites/email/titan-complete-mailbox-setup/index.js]: This looks good, but varies from depictions found in the design post

File: client/my-sites/email/titan-add-mailboxes/titan-new-mailbox-list.jsx
Comment (daledupreez) [reply client/my-sites/email/titan-add-mailboxes/titan-new-mailbox-list.jsx]: Sorry about not explaining my reasoning -- most of my concern was that `supportsMultipleMailboxes` is somewhat passive and is more of a state than an instruction/command for the component. From that perspective, I interpret/understand `allowMultipleMailboxes` to be an instruction for the component to affect/direct its behaviour, much like the `showLabels` prop.
I could see `supportMultipleMailboxes` being a valid name using that naming approach, because we're giving the component an instruction.
But I don't think this naming conversation should block the PR! I'll kick off a dedicated async discussion on this.

File: client/my-sites/email/titan-complete-mailbox-setup/index.js
Comment (stephanethomas) [inline client/my-sites/email/titan-complete-mailbox-setup/index.js]: Each word should be capitalized, i.e. `Complete Setup`.

**Update:** I see now that this was made on purpose in order to reuse existing translations. So another option could be to keep it as it-is, and simply capitalize words [via CSS](https://developer.mozilla.org/en-US/docs/Web/CSS/text-transform).

**Update 2:** after giving it more thoughts, I think we should just show `Set Up` as CTA. It's shorter, and it matches the title of the page (from the `<HeaderCake>` component). It is also future-proof as we won't need to change it when we implement the flow for unused mailboxes.

File: client/my-sites/email/titan-complete-mailbox-setup/index.js
Comment (stephanethomas) [inline client/my-sites/email/titan-complete-mailbox-setup/index.js]: I think `Set up your Professional Email` is enough, and this would actually match the call-to-action [shown](https://github.com/Automattic/wp-calypso/pull/54929/files#r678259957) on the `Customer Home` page.

File: client/my-sites/email/titan-complete-mailbox-setup/index.js
Comment (stephanethomas) [inline client/my-sites/email/titan-complete-mailbox-setup/index.js]: We should use the same title than in the `<HeaderCake>` component below but capitalize all words ([an example](https://github.com/Automattic/wp-calypso/blob/9bee04a60fdd34d64c62b2591c599f3c0d4d6929/client/my-sites/email/email-management/titan-manage-mailboxes/index.jsx#L107)).

File: client/my-sites/email/titan-complete-mailbox-setup/index.js
Comment (stephanethomas) [inline client/my-sites/email/titan-complete-mailbox-setup/index.js]: Apologies if I've missed something but as far I can see `selectedDomainName` will always be available as it is a required prop. Since this function will actually redirect the user to the `Email Plan` page, I would also rename `goToEmail()` to `goToEmailPlan()`.

File: client/my-sites/email/titan-complete-mailbox-setup/index.js
Comment (stephanethomas) [inline client/my-sites/email/titan-complete-mailbox-setup/index.js]: I'm not sure about showing `Email Management` here. First, because we usually show `Back` as text. Second, because when I click that link I'm actually redirected to the `Email Plan` page, not the `Emails` page. And finally, because it's a bit long - something that doesn't help with the already long title on small viewports:

screenshot


I understand the user will access this new page from the [setup card](https://github.com/Automattic/wp-calypso/pull/54929) shown on the `Customer Home` page. What if we simply showed the standard `Back` link, and redirected them to the `Customer Home` if they click it?

File: client/data/emails/use-create-titan-mailbox-mutation.js
Comment (stephanethomas) [inline client/data/emails/use-create-titan-mailbox-mutation.js]: I'm a bit lost: sometimes the type is listed first and other times it is listed after the parameter name.

File: client/my-sites/email/controller.js
Comment (stephanethomas) [inline client/my-sites/email/controller.js]: I think we could shorten the name of those new components and functions:

- `emailManagementCompleteTitanMailboxSetup` → `emailManagementTitanSetupMailbox`
- `TitanCompleteMailboxSetup` → `TitanSetupMailbox`
- `titan-complete-mailbox-setup` →  `titan-setup-mailbox`

This would also match the actual title of the page.

File: client/my-sites/email/paths.js
Comment (stephanethomas) [inline client/my-sites/email/paths.js]: Similarly to my above comment, let's simplify a bit that path. So instead of:

https://wordpress.com/email/:domain/titan/complete-mailbox-setup/:site

We could use:

https://wordpress.com/email/:domain/titan/setup/:site

File: client/my-sites/email/titan-add-mailboxes/titan-new-mailbox-list.jsx
Comment (stephanethomas) [reply client/my-sites/email/titan-add-mailboxes/titan-new-mailbox-list.jsx]: I've already mentioned it internally but I don't have a strong opinion about these two options (I would use `supportMultipleMailboxes` instead of `supportsMultipleMailboxes` though). `showAddAnotherMailboxButton` could be a good alternative that better conveys what that flag actually does.

File: client/my-sites/email/titan-complete-mailbox-setup/index.js
Comment (stephanethomas) [inline client/my-sites/email/titan-complete-mailbox-setup/index.js]: I believe this should be named `getMailboxName` to be consistent with the other parts of the code you're introducing (it doesn't return `mailbox.name.value` anyway).

File: client/my-sites/email/titan-complete-mailbox-setup/index.js
Comment (stephanethomas) [inline client/my-sites/email/titan-complete-mailbox-setup/index.js]: `mailboxIsValid` should be renamed to `isMailboxValid`.

File: client/my-sites/email/titan-complete-mailbox-setup/index.js
Comment (stephanethomas) [inline client/my-sites/email/titan-complete-mailbox-setup/index.js]: We should rename `siteDomainsAreLoaded` to `areSiteDomainsLoaded`.

File: client/my-sites/email/titan-complete-mailbox-setup/index.js
Comment (stephanethomas) [inline client/my-sites/email/titan-complete-mailbox-setup/index.js]: This file is already quite dense, what if we moved `SetupForm` to another file?

File: client/my-sites/email/titan-complete-mailbox-setup/index.js
Comment (stephanethomas) [inline client/my-sites/email/titan-complete-mailbox-setup/index.js]: We could remove that variable, it's only used once and it is almost as long as the function called.

File: client/my-sites/email/titan-complete-mailbox-setup/index.js
Comment (stephanethomas) [inline client/my-sites/email/titan-complete-mailbox-setup/index.js]: Same remark than above regarding the naming of `siteDomainsAreLoaded`.

**Review by stephanethomas — CHANGES_REQUESTED** [review]
Nice work using React Query for this new form 👏.

I left some suggestions that focus on the user experience. I also had comments regarding how to simplify the names of components. Otherwise, my biggest concern is the validation of the form which didn't work correctly for me. Unlike the `Add New Mailboxes` page, clicking on the `Complete setup` button indeed doesn't trigger any error. Actually, no error is shown for the last two fields even if I enter a name and email address. We should investigate and fix that issue.

Finally, I was able to set up an unused mailbox successfully while I thought it failed at first: I was indeed redirected on the `Email Plan` page but no success message was displayed. It took also something like twenty seconds for the list of mailboxes to update and show the new one. We should address that before merging.

File: client/my-sites/email/titan-complete-mailbox-setup/index.js
Comment (olaseni) [reply client/my-sites/email/titan-complete-mailbox-setup/index.js]: I think the opt-in via the sign-up flow indicates an initiation of the setup process, and this page continues the process, hence the need to use the word `Complete`.

I'm totally fine with using what works best, but I just want to ensure that the intent is not lost since we started working on this via design documents. @saygunnyc would you want to weigh in here?

Regarding making it future-proof, I think that's not necessary as we can simply make it a configurable prop for the component.

Comment (olaseni) [reply client/my-sites/email/titan-complete-mailbox-setup/index.js]: Pending when more comments come in though, `Set Up` works.

File: client/my-sites/email/titan-complete-mailbox-setup/index.js
Comment (olaseni) [reply client/my-sites/email/titan-complete-mailbox-setup/index.js]: `siteDomainsAreLoaded` depends on state using a hook selector. It shouldn't be removed.

File: client/my-sites/email/titan-complete-mailbox-setup/index.js
Comment (olaseni) [reply client/my-sites/email/titan-complete-mailbox-setup/index.js]: Yes you are right, but I prefer the readability it offers in this scenario.

File: client/my-sites/email/controller.js
Comment (olaseni) [reply client/my-sites/email/controller.js:68]: In this case I'd rather not reduce specificity

File: client/my-sites/email/titan-complete-mailbox-setup/index.js
Comment (olaseni) [reply client/my-sites/email/titan-complete-mailbox-setup/index.js]: I prefer this specificity as there's also a mailbox property called `name`

File: client/my-sites/email/paths.js
Comment (olaseni) [reply client/my-sites/email/paths.js]: I would prefer: https://wordpress.com/email/:domain/titan/setup-mailbox/:site

File: client/my-sites/email/titan-complete-mailbox-setup/index.js
Comment (stephanethomas) [reply client/my-sites/email/titan-complete-mailbox-setup/index.js]: I'm not saying we should remove it, I'm saying we should rename it from `siteDomainsAreLoaded` to `areSiteDomainsLoaded` 😉.

File: client/my-sites/email/paths.js
Comment (stephanethomas) [reply client/my-sites/email/paths.js]: I don't have strong feelings about it. I'm just slightly concerned that this would not play well if we decide to allow users to set up several mailboxes at once.

Comment (stephanethomas) [reply client/my-sites/email/paths.js]: Following-up on [a discussion](https://github.com/Automattic/wp-calypso/pull/54577#discussion_r679132895) about the url of the `Thank You` page, I realized that the latter and this new Setup page should probably share the same path. With that being said, I'm tempted to favor the shortest option mentioned above so we have those urls:

https://wordpress.com/email/:domain/titan/setup/:site
https://wordpress.com/email/:domain/titan/setup/thank-you/:mailbox/:site

File: client/my-sites/email/titan-complete-mailbox-setup/index.js
Comment (stephanethomas) [reply client/my-sites/email/titan-complete-mailbox-setup/index.js]: Oh, I see this where the confusion comes from now:

https://github.com/Automattic/wp-calypso/blob/4940d44a93cc23a6b293ccac4ddb1c63533924b6/client/lib/titan/new-mailbox.js#L360-L361

**Comment by olaseni** [comment]
> my biggest concern is the validation of the form which didn't work correctly for me. Unlike the `Add New Mailboxes` page, clicking on the `Complete setup` button indeed doesn't trigger any error. Actually, no error is shown for the last two fields even if I enter a name and email address. We should investigate and fix that issue.
> 

Oversight on my part. The issue was that the validation worked but the error messages were not being shown. That has been corrected.

> Finally, I was able to set up an unused mailbox successfully while I thought it failed at first: I was indeed redirected on the `Email Plan` page but no success message was displayed. It took also something like twenty seconds for the list of mailboxes to update and show the new one. We should address that before merging.

The redirect to the `Email Plan` page is temporary. When the `Thank You` page is finished (i.e. it's still in review), successful  actions would be redirected to that page instead. This is akin to how we treat checkouts. This should allow for the whole creation process to complete since it's actually handled asynchronously.

File: client/data/emails/use-create-titan-mailbox-mutation.js
Comment (daledupreez) [inline client/data/emails/use-create-titan-mailbox-mutation.js:28]: Maybe for a follow-up PR at this stage, but it may be much simpler to push all the mailbox arguments into an object that makes it much simpler for us to refactor callers and logic. JSDoc can also support this kind of syntax.
```suggestion
export function useCreateTitanMailboxMutation(
	{
		domainName,
		name,
		mailboxName,
		password,
		alternateEmailAddress,
		isAdmin = false,
	},
	mutationOptions = null /* This could also be pushed into the object, TBH, but I also like it separate as it represents something different from the data/fields */
) {
```

File: client/data/emails/use-get-titan-mailbox-availability.js
Comment (daledupreez) [inline client/data/emails/use-get-titan-mailbox-availability.js:25]: As above, we may want to restructure the data fields into an object to simplify the long-term maintenance for this hook.

File: client/my-sites/email/titan-setup-mailbox/index.js
Comment (daledupreez) [inline client/my-sites/email/titan-setup-mailbox/index.js]: I think we should drop the trailing "mailbox" to match the title in `<DocumentHead>` and `<HeaderCake>` below.

File: client/my-sites/email/titan-setup-mailbox/index.js
Comment (daledupreez) [inline client/my-sites/email/titan-setup-mailbox/index.js:68]: Tiny nit: we could also use `siteId` here. I am also not sure if `selectedSite` could be null, in which case the current code could also throw an error.
```suggestion
			{ selectedSite && <QuerySiteDomains siteId={ siteId } /> }
```

File: client/my-sites/email/titan-setup-mailbox/titan-setup-mailbox-form.jsx
Comment (daledupreez) [inline client/my-sites/email/titan-setup-mailbox/titan-setup-mailbox-form.jsx:41]: I have two non-blocking nit-like items for this function -- we can handle these in a follow-up.

1. It's not immediately clear that this function needs to be wrapped by `dispatch()` - this can complicate calling code.
2. It may be simpler to use a single object argument.

My proposal here handles both of these by requiring `dispatch` in the arguments.
```suggestion
const recordCompleteSetupClickEvent = ( { canContinue, dispatch, mailbox } ) => {
	dispatch(
		recordTracksEvent(
			'calypso_email_management_titan_complete_mailbox_setup_complete_setup_button_click',
			{
				can_continue: canContinue,
				domain_name: getMailboxDomainName( mailbox ),
				mailbox_name: getMailboxUserName( mailbox ),
			}
		)
	);
};
```

File: client/my-sites/email/titan-setup-mailbox/titan-setup-mailbox-form.jsx
Comment (daledupreez) [inline client/my-sites/email/titan-setup-mailbox/titan-setup-mailbox-form.jsx:47]: As noted elsewhere, this function may be easier to maintain long-term if we use destructuring for the arguments.

File: client/my-sites/email/titan-setup-mailbox/titan-setup-mailbox-form.jsx
Comment (daledupreez) [inline client/my-sites/email/titan-setup-mailbox/titan-setup-mailbox-form.jsx]: Nit: this should be `areSiteDomainsLoaded` instead of `siteDomainsAreLoaded`. 😁

File: client/my-sites/email/titan-setup-mailbox/index.js
Comment (daledupreez) [inline client/my-sites/email/titan-setup-mailbox/index.js]: I think we should be forwarding to `emailManagementPurchaseNewEmailAccount()` instead, as that is the page we want to use for upsells.

**Review by daledupreez — CHANGES_REQUESTED** [review]
Overall this is looking really good, but I have two issues we should address:

> Confirm that you are redirect to the Email Home when there's no Titan subscription
We currently redirect to the email plan page -- we should redirect to the purchase page and update the testing instructions.

The second issue is around the behaviour for the mailbox availability check -- I see it firing multiple times for a single submission. If this isn't quick to diagnose, I think we should ship this and tackle that issue in a follow-up PR so we can get the PRs that depend on this one shipped.

File: client/data/emails/use-create-titan-mailbox-mutation.js
Comment (olaseni) [reply client/data/emails/use-create-titan-mailbox-mutation.js:28]: I'll leave this for a follow up, but I generally think it's easier and better for this to be represented and visible in code, rather than just JSDoc.

Save for the last parameter which has a default, they are all required REST API parameters.

File: client/my-sites/email/titan-setup-mailbox/index.js
Comment (olaseni) [reply client/my-sites/email/titan-setup-mailbox/index.js:68]: React will conditionally render the component even if `selectedSite` is null. Also it's pretty hard to find a situation where the site is null.

**Comment by olaseni** [comment]
The issues around the redirect and multiple availability queries have been resolved. We can definitely effect all code quality suggestions in a new iteration.

File: client/data/emails/use-create-titan-mailbox-mutation.js
Comment (daledupreez) [inline client/data/emails/use-create-titan-mailbox-mutation.js:14]: Nit: it looks like some of these lines have additional whitespace after the `@param`.

File: client/my-sites/email/titan-setup-mailbox/index.js
Comment (daledupreez) [inline client/my-sites/email/titan-setup-mailbox/index.js:56]: We should probably redirect to `emailManagement()` if the user has a Google subscription. But we can do that in a follow-up.

**Review by daledupreez — APPROVED** [review]
Tested and LGTM! I'll get this deployed now.

File: client/data/emails/use-create-titan-mailbox-mutation.js
Comment (daledupreez) [reply client/data/emails/use-create-titan-mailbox-mutation.js:14]:
```suggestion
 * @param {string} name The name of the user associated with the mailbox
 * @param {string} mailboxName The mailbox name
 * @param {string} password The password to the mailbox
```

**Comment by a8ci18n** [comment]
This Pull Request is now available for translation here: https://translate.wordpress.com/deliverables/6304414

Thank you @olaseni for including a screenshot in the description! This is really helpful for our translators.

**Comment by a8ci18n** [comment]
Translation for this Pull Request [has now been finished](https://translate.wordpress.com/deliverables/overview/6304414).

**Pull request merged**
